File: UVA/10591_Happy_Number/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    long long cases;
    cin >> cases;

    for (long long i = 1; i < cases + 1; i++) {
        long long num;
        cin >> num;

        long long tmp = num;

        unordered_set<int> record;
        record.insert(tmp);

        while (1) {
            long long square_sum = 0;
            while (tmp > 0) {
                square_sum += pow(tmp % 10, 2);
                tmp /= 10;
            }

            if (square_sum == 1) {
                cout << "Case #" << i << ": " << num << " is a Happy number.\n";
                break;
            } else if (record.count(square_sum) == 1) {
                cout << "Case #" << i << ": " << num << " is an Unhappy number.\n";
                break;
            } else {
                record.insert(square_sum);
            }
        
            tmp = square_sum;
        }
    }

    return 0;
}

File: UVA/10591_Happy_Number/solution.py
import math

cases = int(input())

for i in range(1, cases+1):
    num = int(input())

    tmp = num
    record = {tmp}
    while 1:
        square_sum = 0
        while tmp > 0:
            square_sum += int(math.pow(tmp % 10, 2))
            tmp //= 10

        if square_sum == 1:
            print(f"Case #{i}: {num} is a Happy number.")
            break
        elif square_sum in record:
            print(f"Case #{i}: {num} is an Unhappy number.")
            break
        else:
            record.add(square_sum)

        tmp = square_sum